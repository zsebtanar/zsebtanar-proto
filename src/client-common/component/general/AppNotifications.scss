.notifications {
  width: 100%;
  position: absolute;
  bottom: 3rem;
}
