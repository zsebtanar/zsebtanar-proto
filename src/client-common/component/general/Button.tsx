import { pickBy, startsWith } from 'ramda'
import * as React from 'react'

interface ButtonProps {
  title?: string
  className?: string
  icon?: string
  disabled?: boolean
  loading?: boolean
  primary?: boolean
  secondary?: boolean
  info?: boolean
  submit?: boolean
  tabIndex?: number
  onAction?: (event:MouseEvent) => void
  children?: React.ReactNode
}

const pickArea = pickBy((v, k) => startsWith('area-', k))

export function Button(props: ButtonProps) {
  const onClick = event => {
    if (props.onAction) {
      event.preventDefault()
      props.onAction(event)
    }
  }

  return (
    <button
      type={props.submit ? 'submit' : 'button'}
      className={`btn ${props.className ||
        (props.primary ? 'btn-primary' : props.info ? 'btn-warning' : 'btn-secondary')}`}
      onClick={onClick}
      tabIndex={props.tabIndex}
      title={props.title}
      disabled={props.disabled || props.loading}
      {...pickArea(props)}
    >
      {props.icon && <i className={`fa fa-${props.icon} fa-fw mr-1`} />}
      {props.children}
      {props.loading && <i className="fa fa-cog fa-spin fa-fw" />}
    </button>
  )
}
