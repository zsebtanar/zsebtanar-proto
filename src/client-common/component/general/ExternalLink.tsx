import * as React from 'react'

export function ExternalLink(props: React.AnchorHTMLAttributes<{}>) {
  return <a {...props} target="_blank" rel="noopener noreferrer">{props.children}</a>
}