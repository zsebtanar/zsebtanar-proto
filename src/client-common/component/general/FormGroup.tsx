import * as React from 'react'
import { uid } from '../../util/uuid'
import { ReactElement } from 'react'

interface FormGroupProps {
  label: string
  col?: number
  className?: string
  children?: React.ReactNode
}

export function FormGroup(props: FormGroupProps) {
  const col = props.col || 4
  const colAlt = 12 - col
  const id = `fg-${uid()}`

  return (
    <div className={`form-group row ${props.className || ''}`}>
      <label className={`col-${col} col-form-label`} htmlFor={id}>
        {props.label}
      </label>
      <div className={`col-${colAlt}`}>
        {React.Children.map(props.children, child => React.cloneElement(child as ReactElement<any>, { id }))}
      </div>
    </div>
  )
}
