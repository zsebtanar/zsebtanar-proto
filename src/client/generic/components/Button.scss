.btn {
  &.btn-inline {
    padding: 0;
    vertical-align: bottom;
  }

  .svg-icon {
    height: 1rem;
  }
}
