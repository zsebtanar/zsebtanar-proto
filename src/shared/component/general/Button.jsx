import * as React from 'react'
import { pickBy, startsWith } from 'ramda'

const pickArea = pickBy((v, k) => startsWith('area-', k))

export default function Button (props) {
  const onClick = (event) => {
    if (props.onAction) {
      event.preventDefault()
      props.onAction(event)
    }
  }

  return (<button
    type={props.submit ? 'submit' : 'button'}
    className={`btn ${props.className || (props.primary ? 'btn-primary' : (props.info ?  'btn-warning' : 'btn-secondary'))}`}
    onClick={onClick}
    tabIndex={props.tabIndex}
    title={props.title}
    disabled={props.disabled || props.loading}
    {...pickArea(props)}
  >
    {props.icon && <i className={`fa fa-${props.icon} fa-fw mr-1`}/>}
    {props.children}
    {props.loading && <i className="fa fa-cog fa-spin fa-fw"/>}
  </button>)
}
