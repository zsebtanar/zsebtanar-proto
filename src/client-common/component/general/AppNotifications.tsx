import * as React from 'react'
import { connect } from 'react-redux'
import { Alert } from './Alert'
import { removeNotification } from '../../store/notifications'
import { isNotZeroInteger } from '../../../shared/util/math'

import './AppNotifications.scss'

interface StoreProps {
  notifications: state.Notification[]
}

interface DispatchProps {
  removeNotification: typeof removeNotification
}

const mapStateToProps = (state: state.Root) => ({
  notifications: state.app.notifications.list
})

const mapDispatchToProps = { removeNotification }

///

export const AppNotifications = connect<StoreProps, DispatchProps, {}>(
  mapStateToProps,
  mapDispatchToProps
)(
  class extends React.Component<StoreProps & DispatchProps> {
    render(): React.ReactNode {
      return (
        <div className="notifications" aria-live="polite">
          <div className="col-md-8 col-lg-4 mx-auto">
            {this.props.notifications.map(this.renderNotifications)}
          </div>
        </div>
      )
    }

    private renderNotifications = (item: state.Notification) => {
      const { id, type, message, options } = item
      const onDismiss = isNotZeroInteger(options.timeout) ? undefined : this.dismissNotify(id)
      return (
        <Alert key={id} type={type} onDismiss={onDismiss}>
          {message}
        </Alert>
      )
    }

    private dismissNotify = (id: string) => () => {
      this.props.removeNotification(id)
    }
  }
)
