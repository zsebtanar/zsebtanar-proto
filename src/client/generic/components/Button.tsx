import React from 'react'
import cx from 'classnames'
import {
  Settings as SettingsIcon,
  Icon as IconType,
  AlertTriangle as AlertTriangleIcon,
} from 'react-feather'
import { Icon } from './icons/Icon'

import './Button.scss'
import { ButtonType } from '../../../shared/generic/types'

interface ButtonProps extends Omit<React.HTMLProps<HTMLButtonElement>, 'type' | 'onClick'> {
  btn?: ButtonType
  icon?: IconType
  loading?: boolean
  submit?: boolean
  inline?: boolean
  onAction?: (event: MouseEvent) => void
  small?: boolean
}

export function Button({
  btn,
  onAction,
  tabIndex,
  icon,
  submit,
  className,
  title,
  disabled,
  loading,
  inline,
  children,
  small,
}: ButtonProps): JSX.Element {
  const onClick =
    onAction &&
    ((event) => {
      event.preventDefault()
      onAction?.(event)
    })

  return (
    <button
      type={submit ? 'submit' : 'button'}
      className={cx(className, 'btn', `btn-${btn || 'secondary'}`, {
        'btn-inline': inline,
        'btn-sm': small,
      })}
      onClick={onClick}
      tabIndex={tabIndex}
      title={title}
      aria-label={title}
      disabled={disabled || loading}
    >
      {icon && <Icon icon={icon} />}
      {children}
      {loading && <Icon icon={SettingsIcon} />}
    </button>
  )
}
