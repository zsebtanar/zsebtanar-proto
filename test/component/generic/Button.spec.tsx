import * as React from 'react'
import { Button } from 'client-common/component/general/Button'
import * as renderer from 'react-test-renderer'

describe('Button', () => {
  it('renders correctly the default', () => {
    const tree = renderer.create(<Button>button text</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly the submit', () => {
    const tree = renderer.create(<Button submit>button text</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly the primary', () => {
    const tree = renderer.create(<Button primary>button text</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
