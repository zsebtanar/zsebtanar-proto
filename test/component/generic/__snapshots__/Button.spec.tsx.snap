// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly the default 1`] = `
<button
  className="btn btn-secondary"
  onClick={[Function]}
  type="button"
>
  button text
</button>
`;

exports[`Button renders correctly the primary 1`] = `
<button
  className="btn btn-primary"
  onClick={[Function]}
  type="button"
>
  button text
</button>
`;

exports[`Button renders correctly the submit 1`] = `
<button
  className="btn btn-secondary"
  onClick={[Function]}
  type="submit"
>
  button text
</button>
`;
